tests/unit/test-util/test.artifacts.js

```diff
@@ -55,4 +56,43 @@ describe('prepareArtifactsDir', () => {
     }
   ));
 
+  it('throws error when checking the directory if lacks writing permissions',
+    () => withTempDir(
+    (tmpDir) => {
+      const tmpPath = path.join(tmpDir.path(), 'build');
+      return fs.mkdir(tmpPath, '0622').then(() => {
```

Comment:
As per discussion with @kumar303, we would both prefer to be able to do not inject the `fs` module as a dependency, and to be able to do this we need to test the above scenarios without any sinon spy.

I gave it a try and it looks like (by this test as an example), we can create all the needed tests using the `tmpDir` strategy:

-  "throws an UsageError when directory exists but lacks writing permissions": can be tested by creating the artifactsDir with permissions 555 (it lacks the write permission), something like:
``` 
      const artifactsDir = path.join(tmpDir.path(), 'dir-nowrite');
      return fs.mkdir(artifactsDir, '555').then(() => {
         ...
```

- "throws an UsageError if lacks permissions to create the directory": by creating a parent dir with permissions 555,
```
       const parentDir = path.join(tmpDir.path(), 'dir-nowrite');
       const artifactsDir = path.join(parentDir, 'artifacts');
       return fs.mkdir(parentDir, '555').then(() => {
```

- "throws an UsageError when it lacks permissions to stat the directory": by creating the parent dir with permissions 444 (it lacks the execute permission and nodejs cannot even stat the artifact dir path)
```
       const parentDir = path.join(tmpDir.path(), 'dir-nowrite');
       const artifactsDir = path.join(parentDir, 'artifacts');
       return fs.mkdir(parentDir, '444').then(() => {
```

- "creates the artifacts dir successfully if the parent dir doesn't exist": this should be pretty simple, it should be mostly like the fist test form this file but with an additional directory level (e.g. something like `path.join(tmpDir.path(), 'build', 'subdir')` 

NOTE: on asserting the error message various, try to be specific enough to do not catch one of the other error message by mistake (e.g. `/user lacks permissions/` is not specific enough, and you can use `.*`, e.g. `/myprefix.*mysuffix/`, to match a more specific error message with a regular expression)